crates/rattler_index/src/lib.rs

```diff
@@ -97,117 +127,317 @@ pub fn package_record_from_conda(file: &Path) -> std::io::Result<PackageRecord>
-                tracing::info!("Could not read package record from {:?}", p);
-                continue;
-            };
-            match t {
-                ArchiveType::TarBz2 => repodata
-                    .packages
-                    .insert(file_name.to_string_lossy().to_string(), record),
-                ArchiveType::Conda => repodata
-                    .conda_packages
-                    .insert(file_name.to_string_lossy().to_string(), record),
-            };
+    let repodata_path = format!("{subdir}/repodata.json");
+    let repodata_bytes = serde_json::to_vec(&repodata)?;
+    op.write(&repodata_path, repodata_bytes).await?;
+
+    Ok(())
+}
+
+/// Create a new `repodata.json` for all packages in the channel at the given directory.
+pub async fn index_fs(
+    channel: impl Into<PathBuf>,
+    target_platform: Option<Platform>,
+    force: bool,
+    max_parallel: usize,
+) -> anyhow::Result<()> {
+    let mut config = FsConfig::default();
+    config.root = Some(channel.into().canonicalize()?.to_string_lossy().to_string());
+    index(target_platform, config, force, max_parallel).await
+}
+
+/// Create a new `repodata.json` for all packages in the channel at the given S3 URL.
+#[allow(clippy::too_many_arguments)]
+pub async fn index_s3(
+    channel: Url,
+    region: String,
+    endpoint_url: Url,
+    force_path_style: bool,
+    access_key_id: Option<String>,
+    secret_access_key: Option<String>,
+    session_token: Option<String>,
+    target_platform: Option<Platform>,
+    force: bool,
+    max_parallel: usize,
+) -> anyhow::Result<()> {
+    let mut s3_config = S3Config::default();
+    s3_config.root = Some(channel.path().to_string());
+    s3_config.bucket = channel
+        .host_str()
+        .ok_or(anyhow::anyhow!("No bucket in S3 URL"))?
+        .to_string();
+    s3_config.region = Some(region);
+    s3_config.endpoint = Some(endpoint_url.to_string());
+    s3_config.enable_virtual_host_style = !force_path_style;
+    // Use credentials from the CLI if they are provided.
+    if let (Some(access_key_id), Some(secret_access_key)) = (access_key_id, secret_access_key) {
+        s3_config.secret_access_key = Some(secret_access_key);
+        s3_config.access_key_id = Some(access_key_id);
+        s3_config.session_token = session_token;
+    } else {
+        // If they're not provided, check rattler authentication storage for credentials.
+        let auth_storage = AuthenticationStorage::from_env_and_defaults()?;
+        let auth = auth_storage.get_by_url(channel)?;
+        if let (
+            _,
+            Some(Authentication::S3Credentials {
+                access_key_id,
+                secret_access_key,
+                session_token,
+            }),
+        ) = auth
+        {
+            s3_config.access_key_id = Some(access_key_id);
+            s3_config.secret_access_key = Some(secret_access_key);
+            s3_config.session_token = session_token;
         }
```

Comment:
is it intended that we don't fail when we don't find any credentials? to me, a public s3 bucket doesn't make much sense as then it's just an http server

Reply:
We don't fail because there might still be the option of passing credentials via environment variables (e.g., `AWS_*`). Opendal claims it handles these as well.